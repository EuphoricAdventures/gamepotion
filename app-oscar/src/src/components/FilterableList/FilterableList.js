import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import Input from '../Input/Input'
import List from '../List/List'

const StyledFilterableList = styled.ul`
  .component--input + .component--list {
    margin-top: 0.5rem;
  }
`

class FilterableList extends React.PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      filter: this.props.filter
    }
    this.onUpdateFilter = this.onUpdateFilter.bind(this)
  }

  onUpdateFilter(filter) {
    this.setState({
      filter
    })
  }

  render() {
    const filteredListItems = this.props.children.filter(c => c.props.children.toLowerCase().indexOf(this.state.filter.toLowerCase()) >= 0)
    return (
      <StyledFilterableList className='component--filterable-list'>
        <Input type='text' placeholder='Filter...' value={this.state.filter} onChange={this.onUpdateFilter} />
        <List>
          {filteredListItems}
        </List>
      </StyledFilterableList>
    )
  }
}

FilterableList.propTypes = {
  filter: PropTypes.string
}

FilterableList.defaultProps = {
  filter: ''
}

export default FilterableList
