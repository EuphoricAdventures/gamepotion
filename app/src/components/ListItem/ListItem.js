import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classnames from 'classnames'

import icons from '../../icons'
import { font, colours } from '../../styleAbstractions'

const StyledListItem = styled.button`
  position: relative;
  display: block;
  width: 100%;
  height: calc(2rem + 4px);
  list-style-type: none;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
  border-radius: 4px;
  background-color: transparent;
  transition: border-color 0.2s ease-in-out;
  outline: 0;
  &.selected {
    background-color: #dadfe1;
  }
  &:focus {
    border-color: ${colours.outline};
  }
  > img {
    display: block;
    float: left;
    width: 1.5rem;
    height: 1.5rem;
    margin-top: 0.25rem;
    margin-left: 0.25rem;
    // background-color: red;
  }
  > span {
    display: block;
    float: left;
    height: 2rem;
    line-height: 2rem;
    margin-left: 0.5rem;
    ${font}
    color: ${colours.fore};
    // background-color: green;
  }
  &:hover {
    > .actions {
      visibility: visible;
      opacity: 1;
    }
  }
  > .actions {
    position: absolute;
    top: 0;
    right: 0;
    height: 2rem;
    transition: opacity 0.1s ease-in-out;
    visibility: hidden;
    opacity: 0;
    // background-color: pink;
    img {
      display: block;
      float: left;
      width: 1.5rem;
      height: 1.5rem;
      margin-top: 0.25rem;
      margin-left: 0.25rem;
      background-color: transparent;
      opacity: 0.5;
      border-radius: 4px;
      transition: opacity 0.1s ease-in-out;
      outline: 0;
      // background-color: blue;
      &:focus {
        opacity: 1;
        background-color: ${colours.outline};
      }
      &:hover {
        opacity: 1;
      }
    }
  }
`

const actions = {
  'add': (id, handleOnAction) => (<img title='Add' key='action-add' onClick={(e) => handleOnAction(e, id, 'add')} className='action' src={icons.generic.actions.add} alt='' tabIndex='0' />),
  'rename': (id, handleOnAction) => (<img title='Rename' key='action-rename' onClick={(e) => handleOnAction(e, id, 'rename')} className='action' src={icons.generic.actions.edit} alt='' tabIndex='0' />),
  'delete': (id, handleOnAction) => (<img title='Delete' key='action-delete' onClick={(e) => handleOnAction(e, id, 'delete')} className='action' src={icons.generic.actions.delete} alt='' tabIndex='0' />)
}

const getAction = (onAction, id, name) => {
  const foundAction = actions[name]
  if (typeof foundAction === 'function') {
    return foundAction(id, onAction)
  }
  return null
}

const ListItem = ({ id, icon, selected, actions, children, onChoose, onAction }) => {
  const handleOnAction = (e, id, action) => {
    e.stopPropagation()
    onAction(id, action)
  }
  return (
    <StyledListItem onClick={() => onChoose(id)} className={classnames('component--list-item', {'selected': selected})}>
      <img src={icon} alt='' />
      <span>{children}</span>
      <div className='actions'>
        {actions.map(name => getAction(handleOnAction, id, name))}
      </div>
    </StyledListItem>
  )
}

ListItem.propTypes = {
  id: PropTypes.string.isRequired,
  icon: PropTypes.string,
  selected: PropTypes.bool,
  actions: PropTypes.array,
  onChoose: PropTypes.func,
  onAction: PropTypes.func
}

ListItem.defaultProps = {
  icon: '',
  selected: false,
  actions: [],
  onChoose: () => {},
  onAction: () => {}
}

export default ListItem
