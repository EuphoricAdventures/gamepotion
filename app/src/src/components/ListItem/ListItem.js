import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import classnames from 'classnames'

import icons from '../../icons'

const styleAbstractions = {
  colours: {
    'outline': '#a537fd',
    'fore': '#2e3131',
  },
  font: `
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif;
    font-weight: 400;
    text-size-adjust: 100%;
    -webkit-font-smoothing: antialiased;
  `
}

const StyledListItem = styled.li`
  position: relative;
  display: block;
  width: calc(100% - 4px);
  height: 2rem;
  list-style-type: none;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
  border-radius: 4px;
  background-color: transparent;
  transition: border-color 0.2s ease-in-out;
  outline: 0;
  cursor: default;
  &.selected {
    background-color: #dadfe1;
  }
  &:focus {
    border-color: ${styleAbstractions.colours.outline};
  }
  > img {
    display: block;
    float: left;
    width: 1.5rem;
    height: 1.5rem;
    margin-top: 0.3rem;
    margin-bottom: 0.3rem;
    margin-left: 0.3rem;
    // background-color: red;
  }
  > span {
    display: block;
    float: left;
    height: 1.5rem;
    line-height: 1.5rem;
    margin-top: 0.3rem;
    margin-bottom: 0.3rem;
    margin-left: 0.5rem;
    ${styleAbstractions.font}
    color: ${styleAbstractions.colours.fore};
    // background-color: green;
  }
  &:hover {
    > .actions {
      visibility: visible;
      opacity: 1;
    }
  }
  > .actions {
    position: absolute;
    top: 0;
    right: 0;
    height: 2rem;
    transition: opacity 0.1s ease-in-out;
    visibility: hidden;
    opacity: 0;
    // background-color: pink;
    img {
      display: block;
      float: left;
      width: 1.5rem;
      height: 1.5rem;
      margin-top: 0.25rem;
      margin-left: 0.25rem;
      opacity: 0.5;
      border-radius: 4px;
      transition: opacity 0.1s ease-in-out, background-color 0.1s ease-in-out;
      outline: 0;
      // background-color: blue;
      &:focus {
        opacity: 1;
        background-color: ${styleAbstractions.colours.outline};
      }
      &:hover {
        opacity: 1;
      }
    }
  }
  > .tags {
    position: absolute;
    top: 0;
    right: 0;
    height: 2rem;
    span {
      display: block;
      float: left;
      height: 1.5rem;
      line-height: 1.5rem;
      margin-top: 0.25rem;
      padding-left: 0.5rem;
      padding-right: 0.5rem;
      border-radius: 5000px;
      ${styleAbstractions.font}
      font-weight: 800;
      font-size: 70%;
      text-transform: uppercase;
      color: white;
    }
    span + span {
      margin-left: 0.5rem;
    }
  }
`

const actions = {
  'load': (id, handleOnAction) => (<img title='Load' key='action-load' onClick={(e) => handleOnAction(e, id, 'load')} className='action' src={icons.generic.actions.load} alt='' tabIndex='0' />),
  'add': (id, handleOnAction) => (<img title='Add' key='action-add' onClick={(e) => handleOnAction(e, id, 'add')} className='action' src={icons.generic.actions.add} alt='' tabIndex='0' />),
  'move-up': (id, handleOnAction) => (<img title='Move up' key='action-move-up' onClick={(e) => handleOnAction(e, id, 'move-up')} className='action' src={icons.generic.actions.move.up} alt='' tabIndex='0' />),
  'move-down': (id, handleOnAction) => (<img title='Move down' key='action-move-down' onClick={(e) => handleOnAction(e, id, 'move-down')} className='action' src={icons.generic.actions.move.down} alt='' tabIndex='0' />),
  'rename': (id, handleOnAction) => (<img title='Rename' key='action-rename' onClick={(e) => handleOnAction(e, id, 'rename')} className='action' src={icons.generic.actions.edit} alt='' tabIndex='0' />),
  'edit': (id, handleOnAction) => (<img title='Edit' key='action-edit' onClick={(e) => handleOnAction(e, id, 'edit')} className='action' src={icons.generic.actions.edit} alt='' tabIndex='0' />),
  'delete': (id, handleOnAction) => (<img title='Delete' key='action-delete' onClick={(e) => handleOnAction(e, id, 'delete')} className='action' src={icons.generic.actions.delete} alt='' tabIndex='0' />)
}

const getAction = (onAction, id, name) => {
  const foundAction = actions[name]
  if (typeof foundAction === 'function') {
    return foundAction(id, onAction)
  }
  return null
}

const ListItem = ({ id, icon, selected, actions, tags, children, onChoose, onAction, indentation }) => {
  const handleKeypress = (e) => {
    if (e.key === 'Enter') {
      onChoose(id)
    }
  }
  const handleOnAction = (e, id, action) => {
    e.stopPropagation()
    onAction(id, action)
  }
  const imgStyle = {
    marginLeft: `${indentation}px`
  }
  return (
    <StyledListItem onClick={() => onChoose(id)} onTouchEnd={() => onChoose(id)} className={classnames('component--list-item', {'selected': selected})} tabIndex='0' onKeyPress={handleKeypress}>
      <img src={icon} style={imgStyle} alt='' />
      <span>{children}</span>
      <div className='actions'>
        {actions.map(name => getAction(handleOnAction, id, name))}
      </div>
      <div className='tags'>
        {tags.map(tag => (
          <span key={tag.name} style={{backgroundColor: tag.colour}}>{tag.name}</span>
        ))}
      </div>
    </StyledListItem>
  )
}

ListItem.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  icon: PropTypes.string,
  selected: PropTypes.bool,
  actions: PropTypes.array,
  tags: PropTypes.arrayOf(PropTypes.object),
  onChoose: PropTypes.func,
  onAction: PropTypes.func,
  indentation: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}

ListItem.defaultProps = {
  icon: '',
  selected: false,
  actions: [],
  tags: [],
  onChoose: () => {},
  onAction: () => {},
  indentation: 4
}

export default ListItem
