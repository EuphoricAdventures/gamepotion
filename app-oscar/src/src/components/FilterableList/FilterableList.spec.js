import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import FilterableList from './FilterableList'
import ListItem from '../ListItem/ListItem'

jest.mock('uuid', () => jest.fn().mockReturnValue('id'))

it('renders', () => {
  const tree = renderer
    .create(
      <FilterableList>
        <ListItem id='list-item-1'>List Item 1</ListItem>
        <ListItem id='list-item-2'>List Item 2</ListItem>
      </FilterableList>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
