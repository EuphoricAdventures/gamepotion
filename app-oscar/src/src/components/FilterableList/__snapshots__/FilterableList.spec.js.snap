// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c1 label {
  display: block;
  width: 100%;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  font-size: 80%;
  color: #6c7a89;
}

.c1 label[aria-disabled="true"] {
  opacity: 0.5;
}

.c1 input {
  display: block;
  width: 100%;
  padding: 0.5rem 0 0.5rem 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: 0;
  border: 0;
  border-radius: 0;
  border-bottom: 2px solid #dadfe1;
  background-color: transparent;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-transition: border-color 0.2s ease-in-out;
  transition: border-color 0.2s ease-in-out;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  color: #2e3131;
  -webkit-appearance: none;
}

.c1 input:focus {
  border-color: #bdc3c7;
}

.c1 input::selection {
  background-color: #dadfe1;
}

.c1 input[disabled] {
  opacity: 0.5;
}

.c1 input::-webkit-input-placeholder {
  color: #bdc3c7;
}

input.c1::-moz-placeholder {
  color: #bdc3c7;
}

.c1 input:-moz-placeholder {
  color: #bdc3c7;
}

.c1 input:-ms-input-placeholder {
  color: #bdc3c7;
}

.c2 .component--list {
  margin-left: 1.5rem;
}

.c2 .component--list .component--list-item:first-of-type {
  margin-top: 0.25rem;
}

.c2 .component--list + .component--list-item {
  margin-top: 0.25rem;
}

.c2 .component--list-item + .component--list-item {
  margin-top: 0.25rem;
}

.c2 .empty-text {
  padding-top: 2rem;
  padding-bottom: 2rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  text-align: center;
  color: #2e3131;
  opacity: 0.5;
}

.c0 .component--input + .component--list {
  margin-top: 0.5rem;
}

.c3 {
  position: relative;
  display: block;
  width: 100%;
  height: 2rem;
  list-style-type: none;
  border-width: 2px;
  border-style: solid;
  border-color: transparent;
  border-radius: 4px;
  background-color: transparent;
  -webkit-transition: border-color 0.2s ease-in-out;
  transition: border-color 0.2s ease-in-out;
  outline: 0;
  cursor: default;
}

.c3.selected {
  background-color: #dadfe1;
}

.c3:focus {
  border-color: #38EF7D;
}

.c3 > img {
  display: block;
  float: left;
  width: 1.5rem;
  height: 1.5rem;
  margin-top: 0.3rem;
  margin-bottom: 0.3rem;
  margin-left: 0.3rem;
}

.c3 > span {
  display: block;
  float: left;
  height: 1.5rem;
  line-height: 1.5rem;
  margin-top: 0.3rem;
  margin-bottom: 0.3rem;
  margin-left: 0.5rem;
  font-family: -apple-system,BlinkMacSystemFont,"Segoe UI",sans-serif;
  font-weight: 400;
  -webkit-text-size-adjust: 100%;
  text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  color: #2e3131;
}

.c3:hover > .actions {
  visibility: visible;
  opacity: 1;
}

.c3 > .actions {
  position: absolute;
  top: 0;
  right: 0;
  height: 2rem;
  -webkit-transition: opacity 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out;
  visibility: hidden;
  opacity: 0;
}

.c3 > .actions img {
  display: block;
  float: left;
  width: 1.5rem;
  height: 1.5rem;
  margin-top: 0.25rem;
  margin-left: 0.25rem;
  opacity: 0.5;
  border-radius: 4px;
  -webkit-transition: opacity 0.1s ease-in-out,background-color 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out,background-color 0.1s ease-in-out;
  outline: 0;
}

.c3 > .actions img:focus {
  opacity: 1;
  background-color: #38EF7D;
}

.c3 > .actions img:hover {
  opacity: 1;
}

<ul
  className="component--list c0"
>
  <div
    className="component--input c1"
  >
    <input
      disabled={false}
      id="component-Input-e394c125-ba91-4ce6-b05e-7abc99bcd4b0"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Filter..."
      required={false}
      type="text"
      value=""
    />
  </div>
  <ul
    className="component--list c2"
  >
    <li
      className="component--list-item c3"
      onClick={[Function]}
      onKeyPress={[Function]}
      onTouchEnd={[Function]}
      tabIndex="0"
    >
      <img
        alt=""
        src=""
        style={
          Object {
            "marginLeft": "4px",
          }
        }
      />
      <span>
        List Item 1
      </span>
      <div
        className="actions"
      />
    </li>
    <li
      className="component--list-item c3"
      onClick={[Function]}
      onKeyPress={[Function]}
      onTouchEnd={[Function]}
      tabIndex="0"
    >
      <img
        alt=""
        src=""
        style={
          Object {
            "marginLeft": "4px",
          }
        }
      />
      <span>
        List Item 2
      </span>
      <div
        className="actions"
      />
    </li>
  </ul>
</ul>
`;
